### PR TITLE
support PyPI trusted publishers

### LINKED ISSUE #17: support Trusted Publishers
- `password.required: false`
- `registry-domain.default: https://upload.pypi.org`
- if `password` undefined, generate OIDC "password"
  ```bash
  audience=$(GET $INPUT_REGISTRY_DOMAIN/_/oidc/audience)
  oidc_token=$(GET $ACTIONS_ID_TOKEN_REQUEST_URL -d "$audience" -H "Authorization: bearer $ACTIONS_ID_TOKEN_REQUEST_TOKEN" | jq "{token: .value}")
  TWINE_PASSWORD=$(POST $INPUT_REGISTRY_DOMAIN/_/oidc/github/mint-token -d "$oidc_token" | jq -r .token)
  ```

<details><summary>references</summary>

- description: https://blog.pypi.org/posts/2023-04-20-introducing-trusted-publishers#using-trusted-publishing-with-github-actions
- implementation: https://github.com/pypa/gh-action-pypi-publish/blob/unstable/v1/oidc-exchange.py
- `detect_github(audience) -> oidc_token`: https://github.com/di/id/blob/main/id/_internal/oidc/ambient.py
- GHA env vars: https://docs.github.com/en/actions/learn-github-actions/variables#default-environment-variables

</details>

### PR DESCRIPTION
**Pull request by casperdcl**
Fixes #17

/CC @0x2b3bfa0

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -121,6 +125,21 @@ runs:
             TWINE_OPTS="$TWINE_OPTS --skip-existing"
           fi
           python -m pip install twine
+          if [[ -z "$TWINE_PASSWORD" ]]; then
+            # vis https://docs.pypi.org/trusted-publishers/using-a-publisher/#the-manual-way
+            export TWINE_PASSWORD=$(
+        python -c 'import json, urllib.request as req
+        audience = req.urlopen("'${INPUT_REGISTRY_DOMAIN}'/_/oidc/audience").read()
```
Comment (0x2b3bfa0) [inline action.yml]: f-string and native `getenv` rather than shell variables, please?

Comment (casperdcl) [reply action.yml]: Why bother with the additional code?

File: action.yml [hunk 2]
```diff
@@ -121,6 +125,21 @@ runs:
...
+        token = json.load(req.urlopen(req.Request(
+            url="'${ACTIONS_ID_TOKEN_REQUEST_URL}'/_/oidc/audience",
```
Comment (0x2b3bfa0) [inline action.yml]: That appended path (/…/audience) shall be deleted.

File: action.yml [hunk 3]
```diff
@@ -121,6 +125,21 @@ runs:
...
+            data=audience,
+            headers={"Authorization": "bearer '${ACTIONS_ID_TOKEN_REQUEST_TOKEN}'"})))
+        print(json.load(req.urlopen(req.Request(
+            url="https://'${INPUT_REGISTRY_DOMAIN}'/_/oidc/github/mint-token",
+            data=json.dumps({"token": token["value"]}).encode("U8"),
```
Comment (0x2b3bfa0) [inline action.yml]: Consider calling `.encode()` sans any parameters; default is UTF-8.

File: action.yml (hunk 2 above)
Comment (casperdcl) [reply action.yml:134]: Future tense? As in this is fine as-is for now?

File: action.yml (hunk 3 above)
Comment (casperdcl) [reply action.yml:139]:
```suggestion
            data=json.dumps({"token": token["value"]}).encode(),
```

File: action.yml [hunk 4]
```diff
@@ -121,6 +125,21 @@ runs:
...
+            method="POST")))["token"], end="")'
+        )
```
Comment (casperdcl) [inline action.yml:141]:
```suggestion
            method="POST")))["token"], end="")')
```

File: action.yml (hunk 1 above)
Comment (0x2b3bfa0) [reply action.yml]: Exceedingly defensive; primed by https://github.com/trufflesecurity/trufflehog/pull/2259

Comment (casperdcl) [reply action.yml]: I don't see a problem...

```sh
$ foo="let's try"
$ echo 'some code "using '${foo}' seems to work" right?'
some code "using let's try seems to work" right?
```

Comment (casperdcl) [reply action.yml]: oh I see

```sh
$ bar='let"s try'
$ python -c 'print("using '${bar}' seems to work")'
SyntaxError
```

File: action.yml [hunk 5]
```diff
@@ -121,6 +125,20 @@ runs:
...
+            # vis https://docs.pypi.org/trusted-publishers/using-a-publisher/#the-manual-way
+            export TWINE_PASSWORD=$(
+        python -c 'import json, urllib.request as req
+        audience = req.urlopen("'${INPUT_REGISTRY_DOMAIN}'/_/oidc/audience").read()
+        token = json.load(req.urlopen(req.Request(
+            url="'${ACTIONS_ID_TOKEN_REQUEST_URL}'",
+            data=audience,
+            headers={"Authorization": "bearer '${ACTIONS_ID_TOKEN_REQUEST_TOKEN}'"})))
+        print(json.load(req.urlopen(req.Request(
+            url="https://'${INPUT_REGISTRY_DOMAIN}'/_/oidc/github/mint-token",
```
Comment (casperdcl) [inline action.yml]: couldn't see an easy way to use f-strings without nested quote hell. wdyt?

```suggestion
        audience = req.urlopen(os.getenv("INPUT_REGISTRY_DOMAIN") + "/_/oidc/audience").read()
        token = json.load(req.urlopen(req.Request(
            url=os.getenv("ACTIONS_ID_TOKEN_REQUEST_URL"),
            data=audience,
            headers={"Authorization": "bearer " + os.getenv("ACTIONS_ID_TOKEN_REQUEST_TOKEN")})))
        print(json.load(req.urlopen(req.Request(
            url=os.getenv("INPUT_REGISTRY_DOMAIN") + "/_/oidc/github/mint-token",
```

(P.S. also fixes fixes duplicate `https://https://`)

**Review by casperdcl — COMMENTED** [review]
looks like [it isn't working](https://github.com/tqdm/py-make/actions/runs/7802896706/job/21281646705)

File: action.yml [hunk 6]
```diff
@@ -121,6 +125,21 @@ runs:
...
+        audience = req.urlopen(os.getenv("INPUT_REGISTRY_DOMAIN") + "/_/oidc/audience").read()
+        token = json.load(req.urlopen(req.Request(
+            url=os.getenv("ACTIONS_ID_TOKEN_REQUEST_URL"),
+            data=audience,
+            headers={"Authorization": "bearer " + os.getenv("ACTIONS_ID_TOKEN_REQUEST_TOKEN")},
+            method="GET")))
+        print(json.load(req.urlopen(req.Request(
+            url=os.getenv("INPUT_REGISTRY_DOMAIN") + "/_/oidc/github/mint-token",
+            data=json.dumps({"token": token["value"]}).encode(),
+            method="POST")))["token"], end="")')
```
Comment (casperdcl) [inline action.yml]: getting `urllib.error.HTTPError: HTTP Error 422: Unprocessable Entity` with this

**Pull request merged**